File: Java/opentraceability/src/main/java/opentraceability/models/events/kdes/EventKDEObject.java
package opentraceability.models.events.kdes;

import opentraceability.Constants;
import opentraceability.interfaces.IEventKDE;
import opentraceability.utility.XElement;
import opentraceability.utility.XmlToJsonConverter;
import org.apache.commons.lang3.NotImplementedException;
import org.json.JSONArray;
import org.json.JSONObject;
import org.json.XML;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.StringReader;

import org.xml.sax.InputSource;

public class EventKDEObject extends IEventKDE {
    public Class valueType = Object.class;
    public Object value = null;

    private XElement _xml = null;
    private JSONObject _json = null;
    private Boolean _isJsonArray = false;

    public EventKDEObject() {}

    public EventKDEObject(String ns, String name) {
        this.namespace = ns;
        this.name = name;
    }

    public Object getJson() {
        if (_xml != null)
        {
            JSONObject j = XmlToJsonConverter.toJSON(_xml);
            return j;
        }
        else if (_json != null)
        {
            if (_isJsonArray)
            {
                return _json.optJSONArray(name);
            }
            else
            {
                return _json;
            }
        } else {
            return null;
        }
    }


    public XElement getXml() {
        if (_xml != null) {
            return _xml;
        }
        else if (_json != null)
        {
            XElement x = XmlToJsonConverter.toXML(_json);
            return x;
        }
        else
        {
            return null;
        }
    }

    public void setFromJson(Object json) throws Exception {
        _xml = null;
        _isJsonArray = false;
        if (json instanceof JSONObject)
        {
            _json = (JSONObject)json;
        }
        else if (json instanceof JSONArray)
        {
            _json = new JSONObject();
            _json.put(name, (JSONArray)json);
            _isJsonArray = true;
        }
        else throw new Exception("_json is not JSONObject");
    }

    public void setFromXml(XElement xml) {
        _isJsonArray = false;
        _xml = xml;
        _json = null;
    }

    public String toString() {
        if (value != null) {
            return value.toString();
        } else {
            return "";
        }
    }
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/models/events/kdes/eventkdeobject.kt
package opentraceability.models.events.kdes

import opentraceability.interfaces.IEventKDE
import opentraceability.Constants
import java.lang.reflect.Type
import org.w3c.dom.*
import javax.xml.parsers.DocumentBuilderFactory
import java.io.StringReader
import org.json.*
import javax.xml.parsers.DocumentBuilder
import org.xml.sax.InputSource
import kotlin.reflect.KType
import kotlin.reflect.full.createType

class EventKDEObject: EventKDEBase, IEventKDE {

    override var valueType: KType = Object::class.createType()

    var value: Object? = null


    var _xml: Element? = null
    var _json: JSONObject? = null


    constructor(ns: String, name: String) {
        this.namespace = ns;
        this.name = name;
    }


    override fun getJson(): JSONObject? {
        if ( _xml != null   ){
            val factory: DocumentBuilderFactory = DocumentBuilderFactory.newInstance()
            val builder: DocumentBuilder = factory.newDocumentBuilder()
            val xmlDoc: Document = builder.parse(InputSource(StringReader(_xml.toString())))

            val json = XML.toJSONObject(xmlDoc.toString())
            return JSONObject(json.toString())
        }
        else if (_json != null)
        {
            return _json;
        }
        else
        {
            return null;
        }
    }




    override fun getXml(): Element? {
        if (_xml != null) {
            return _xml
        } else if (_json != null) {

            val xmlDoc = XML.toJSONObject(_json.toString())
            val xmlStr = xmlDoc.toString()

            if (!xmlStr.isNullOrEmpty()) {
                val factory: DocumentBuilderFactory = DocumentBuilderFactory.newInstance()
                val builder: DocumentBuilder = factory.newDocumentBuilder()
                val document: Document = builder.parse(InputSource(StringReader(xmlStr)))

                return document.documentElement
            }

            return null
        } else {
            return null
        }
    }

    override fun setFromJson(json: JSONObject) {
        _xml = null
        _json = json
    }

    override fun setFromXml(xml: Element) {
        _xml = xml
        _json = null
    }

    override fun toString(): String {
        if (value != null) {
            return value.toString()
        } else {
            return ""
        }
    }
}
